src/internal/m365/backup.go

```diff
@@ -186,3 +188,56 @@ func verifyBackupInputs(sels selectors.Selector, cachedIDs []string) error {
 
 	return nil
 }
+
+func (ctrl *Controller) CollectMetadata(
```

Comment:
This does not really have to be a method on `Controller`.

Reply:
The method design is appropriate in this case, actually, since it allows us to pass the controller into the metadata generation as an interface.  However, the larger function design has issues (see my other comment).

The only code we need here are lines 207-211 (and whatever else to transform them to a restore collection).

Reply:
I like the idea of returning paths.  That way we can centralize all metadata path -> collection generation within the operations layer, further removing it from provider/service ownership.  Better control over how we generate those collections that way.

Reply:
I like the idea of returning a list of paths from here. I have to however point out that in case of Groups SharePoint libraries, we first read the `.../libraries/previouspaths` file to figure out what sites are available in `groups.MetadataFiles` which will construct and read the metadata file for which we will need kopia to be passed into here.

Reply:
that kind of makes it feel like having additional nesting to get the metadata is probably not the best choice in the long run. Having to pass kopia directly into the graph connector is going to further complicate already complex logic and lead to more possible failure scenarios

Are there viable alternatives to the layout that's in use at the moment? How much effort would be required to implement one of the alternative?

Reply:
alright, sorry, paged back in some context for this. To make sure I have the state of the world right, what we have for metadata is something like
```
- tenantID
  - GroupsMetadata
    - siteIDList
    - libraries
      - siteID1
        - previouspaths
        - previousdeltas
      - siteID2
        - previouspaths
        - previousdeltas
```

And it's the part where we need to know the siteID that's causing us problems right? We've essentially made a circular dependency in that in order to know the path for the metadata files for a site we need to know the ID of the site which is itself stored in a metadata file. We're also not currently resolving this circular dependency in an alternate way by, for example, enumerating the set of sites prior to asking for metadata (in which case the set of sites that we need to pull metadata for would be known and the siteIDList file would give us the set of deleted sites. Sites that have been deleted shouldn't require metadata from corso because we can just send a single collection at the site level marking the entire subtree deleted)

So given the above, the long-term options appear to be
1. keep what we have laid out re: layout and base selection
2. change layout so we don't need to know additional state to get all required metadata (i.e. collapse metadata into top-level files, store in more generic location, etc. The key is to pick some location that doesn't require additional information about the data we are/have backed up previously in order to find)
3. add more advanced logic in kopia to do wildcard and/or recursive matches so we just do something like "return me all collections with data under this prefix" though it'd get tricker to define what the restore path for those collections should be in the generic case
4. lookup the sites being backed up with graph connector prior to fetching metadata. This would remove the circular dependency because we'd know the paths we need to fetch metadata for. We should ignore errors if things aren't found for a site (i.e. a new site was added since the last backup). Other logic should handle if we're only missing one of the metadata files (i.e. have previous deltas but not previous paths)

Reply:
> We're also not currently resolving this circular dependency in an alternate way by, for example, enumerating the set of sites prior to asking for metadata (in which case the set of sites that we need to pull metadata for would be known and the siteIDList file would give us the set of deleted sites. Sites that have been deleted shouldn't require metadata from corso because we can just send a single collection at the site level marking the entire subtree deleted)

This sounds like a viable solution to me. We could implement this without having to make any major changes IIUC. The only trouble if anything would be is how we manage errors and making sure we are erroring out because it is a new site and not because we did not have it in kopia.

Reply:
>The only trouble if anything would be is how we manage errors and making sure we are erroring out because it is a new site and not because we did not have it in kopia.

IIRC, we already use a separate fault bus when fetching metadata so the errors from there shouldn't cause the backup to fail overall. Not finding metadata for a new site should be something we can ignore. It's new, we'll just do a full enumeration of the data in the site. Not finding metadata for a site we previously backed up should be manageable. We should mark all collections in it as new/do not merge and delete the old hierarchy that existed for the site (so we don't accidentally pull forward folders that were deleted)